evm/src/cpu/kernel/optimizer.rs

```diff
@@ -0,0 +1,135 @@
+use ethereum_types::U256;
+use Item::{Push, StandardOp};
+use PushTarget::Literal;
+
+use crate::cpu::kernel::ast::Item::{GlobalLabelDeclaration, LocalLabelDeclaration};
+use crate::cpu::kernel::ast::PushTarget::Label;
+use crate::cpu::kernel::ast::{Item, PushTarget};
+use crate::cpu::kernel::utils::replace_windows;
+
+pub(crate) fn optimize_asm(code: &mut Vec<Item>) {
+    // Run the optimizer until nothing changes.
+    loop {
+        let old_code = code.clone();
+        optimize_asm_once(code);
+        if code == &old_code {
+            break;
+        }
+    }
+}
+
+/// A single optimization pass.
+fn optimize_asm_once(code: &mut Vec<Item>) {
+    constant_propagation(code);
+    no_op_jumps(code);
+    remove_swaps(code);
+}
+
+/// Constant propagation.
+fn constant_propagation(code: &mut Vec<Item>) {
+    // Constant propagation for unary ops: `[PUSH x, UNARYOP] -> [PUSH UNARYOP(x)]`
+    replace_windows(code, |window| {
+        if let [Push(Literal(x)), StandardOp(op)] = window {
+            match op.as_str() {
+                "ISZERO" => Some(vec![Push(Literal(if x.is_zero() {
+                    U256::one()
+                } else {
+                    U256::zero()
+                }))]),
+                "NOT" => Some(vec![Push(Literal(!x))]),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    });
+
+    // Constant propagation for binary ops: `[PUSH x, PUSH y, BINOP] -> [PUSH BINOP(x, y)]`
+    replace_windows(code, |window| {
+        if let [Push(Literal(x)), Push(Literal(y)), StandardOp(op)] = window {
+            match op.as_str() {
+                "ADD" => Some(vec![Push(Literal(x + y))]),
+                "SUB" => Some(vec![Push(Literal(x - y))]),
+                "MUL" => Some(vec![Push(Literal(x * y))]),
+                "DIV" => Some(vec![Push(Literal(x / y))]),
```

Comment:
Does this have correct edge-case behavior? Rust generally panics or has UB when an integer operation overflows, whereas the EVM wraps around. `wrapping_add`/`sub`/`mul` may be more appropriate. Similarly, division by `0` should produce `0` and not panic.

Reply:
Good point! For some reason I thought `U256` implemented "EVM semantics" rather than "Rust semantics", but it seems that's not the case. I made the changes and added a few tests.